File: ios-app/SocietyManager/Services/AuthService.swift
import Foundation

final class AuthService {
    static let shared = AuthService()
    private let api = APIClient.shared
    private init() {}

    func login(email: String, password: String) async throws -> LoginResponse {
        let request = LoginRequest(email: email, password: password)
        return try await api.request(AppConstants.API.Auth.login, method: .post, body: request)
    }

    func register(name: String, email: String, password: String, role: String) async throws -> User {
        let request = RegisterRequest(name: name, email: email, password: password, role: role)
        return try await api.request(AppConstants.API.Auth.register, method: .post, body: request)
    }

    func logout() async throws {
        try await api.requestVoid(AppConstants.API.Auth.logout, method: .post)
    }

    func getCurrentUser() async throws -> CurrentUser {
        return try await api.request(AppConstants.API.Auth.me)
    }

    func forgotPassword(email: String) async throws {
        let request = ForgotPasswordRequest(email: email)
        try await api.requestVoid(AppConstants.API.Auth.forgotPassword, method: .post, body: request)
    }

    func resetPassword(token: String, newPassword: String) async throws {
        let request = ResetPasswordRequest(token: token, newPassword: newPassword)
        try await api.requestVoid(AppConstants.API.Auth.resetPassword, method: .post, body: request)
    }

    func changePassword(currentPassword: String, newPassword: String) async throws {
        let request = ChangePasswordRequest(currentPassword: currentPassword, newPassword: newPassword)
        try await api.requestVoid(AppConstants.API.Auth.changePassword, method: .post, body: request)
    }
}

File: backend/src/main/java/com/society/backend/auth/service/AuthService.java
package com.society.backend.auth.service;

import com.society.backend.auth.dto.request.LoginRequest;
import com.society.backend.auth.dto.request.LogoutRequest;
import com.society.backend.auth.dto.response.LoginResponse;
import com.society.backend.auth.dto.request.RegisterRequest;
import com.society.backend.user.dto.response.UserResponse;

import jakarta.servlet.http.HttpServletRequest;

public interface AuthService {

    UserResponse register(RegisterRequest request);

    LoginResponse login(LoginRequest request, HttpServletRequest httpRequest);

    void recordLogout(String token, HttpServletRequest httpRequest, LogoutRequest logoutRequest);

    UserResponse getUserFromToken(String token);

    void forgotPassword(String email);

    void resetPassword(String token, String newPassword);

    void changePassword(Long userId, String currentPassword, String newPassword, String specialKey);

    /** Updates the most-recent LOGIN audit's lat/lng for a SOCIETY_ADMIN session. */
    void updateCurrentLocation(Long userId, Double latitude, Double longitude);
}
